library/java/org/chromium/net/AndroidNetworkLibrary.java

```diff
@@ -83,4 +103,153 @@ public static void clearTestRootCertificates()
+      }
+      protected OutputStream getOutputStream() {
+        throw new RuntimeException("getOutputStream not implemented");
+      }
+      protected void listen(int backlog) { throw new RuntimeException("listen not implemented"); }
+      protected void sendUrgentData(int data) {
+        throw new RuntimeException("sendUrgentData not implemented");
+      }
+      public Object getOption(int optID) {
+        throw new RuntimeException("getOption not implemented");
+      }
+      public void setOption(int optID, Object value) {
+        throw new RuntimeException("setOption not implemented");
+      }
+    }
+
+    /**
+     * Create a {@link java.net.Socket} that sets {@code fd} as the underlying file
+     * descriptor. Does not take ownership of the file descriptor, so calling {@link #close}
+     * will not actually close the file descriptor.
+     */
+    SocketFd(FileDescriptor fd) throws IOException { super(new SocketImplFd(fd)); }
+  }
+
+  /**
+   * Tag socket referenced by {@code ifd} with {@code tag} for UID {@code uid}.
+   *
+   * Assumes thread UID tag isn't set upon entry, and ensures thread UID tag isn't set upon exit.
+   * Unfortunately there is no TrafficStatis.getThreadStatsUid().
+   */
+  private static void tagSocket(int ifd, int uid, int tag) throws IOException {
+    // Set thread tags.
+    int oldTag = TrafficStats.getThreadStatsTag();
+    if (tag != oldTag) {
+      TrafficStats.setThreadStatsTag(tag);
+    }
+    // if (uid != TrafficStatsUid.UNSET) {
+    if (uid != -1) {
+      ThreadStatsUid.set(uid);
+    }
+
+    // Apply thread tags to socket.
+
+    // First, convert integer file descriptor (ifd) to FileDescriptor.
+    final ParcelFileDescriptor pfd;
+    final FileDescriptor fd;
+    // The only supported way to generate a FileDescriptor from an integer file
+    // descriptor is via ParcelFileDescriptor.adoptFd(). Unfortunately prior to Android
+    // Marshmallow ParcelFileDescriptor.detachFd() didn't actually detach from the
+    // FileDescriptor, so use reflection to set {@code fd} into the FileDescriptor for
+    // versions prior to Marshmallow. Here's the fix that went into Marshmallow:
+    // https://android.googlesource.com/platform/frameworks/base/+/b30ad6f
+    if (Build.VERSION.SDK_INT < VERSION_CODES.M) {
+      pfd = null;
+      fd = SetFileDescriptor.createWithFd(ifd);
+    } else {
+      pfd = ParcelFileDescriptor.adoptFd(ifd);
+      fd = pfd.getFileDescriptor();
+    }
+    // Second, convert FileDescriptor to Socket.
+    Socket s = new SocketFd(fd);
+    // Third, tag the Socket.
+    TrafficStats.tagSocket(s);
+    s.close(); // No-op but always good to close() Closeables.
+    // Have ParcelFileDescriptor relinquish ownership of the file descriptor.
+    if (pfd != null) {
+      pfd.detachFd();
+    }
+
+    // Restore prior thread tags.
+    if (tag != oldTag) {
+      TrafficStats.setThreadStatsTag(oldTag);
+    }
+    //    if (uid != TrafficStatsUid.UNSET) {
```

Comment:
?

Reply:
Ah, removed. Thanks. (In cronet that enum comes from some C code that we're not building in Envoy. )